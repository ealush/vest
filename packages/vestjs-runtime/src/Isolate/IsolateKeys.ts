export enum IsolateKeys {
  Type = '$type',
  Keys = 'keys',
  Key = 'key',
  Parent = 'parent',
  Data = 'data',
  AllowReorder = 'allowReorder',
  Status = 'status',
}

enum MinifiedKeys {
  Type = '$',
  Keys = 'K',
  Key = 'k',
  Parent = 'P',
  Data = 'D',
  AllowReorder = 'aR',
  Status = 'S',
}

export const KeyToMinified = {
  [IsolateKeys.Type]: MinifiedKeys.Type,
  [IsolateKeys.Keys]: MinifiedKeys.Keys,
  [IsolateKeys.Parent]: MinifiedKeys.Parent,
  [IsolateKeys.Data]: MinifiedKeys.Data,
  [IsolateKeys.Key]: MinifiedKeys.Key,
  [IsolateKeys.AllowReorder]: MinifiedKeys.AllowReorder,
  [IsolateKeys.Status]: MinifiedKeys.Status,
};

// This const is an object that looks like this:
// {
//   '$': '$type',
//   'K': 'keys',
//   'P': 'parent',
//   ...
// }
export const MinifiedToKey = Object.entries(KeyToMinified).reduce(
  (acc, [key, minified]) =>
    Object.assign(acc, {
      [minified]: key,
    }),
  {} as Record<string, IsolateKeys>
);
