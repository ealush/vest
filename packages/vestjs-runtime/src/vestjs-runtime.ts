export { IsolateKey, TIsolate } from 'Isolate';
export { Reconciler, IRecociler } from 'Reconciler';
export * as Walker from 'IsolateWalker';
export { RuntimeApi as VestRuntime } from 'VestRuntime';
export { IsolateInspector } from 'IsolateInspector';
export { IsolateMutator } from 'IsolateMutator';
export * as Bus from 'Bus';
export * as isolateInterfaces from 'isolateInterfaces';
export * as IsolateSelectors from 'IsolateSelectors';
