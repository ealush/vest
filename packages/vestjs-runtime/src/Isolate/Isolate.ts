import { CB, Nullable } from 'vest-utils';

import { IsolateKeys } from 'IsolateKeys';
import { IsolateMutator } from 'IsolateMutator';
import { Reconciler } from 'Reconciler';
import * as VestRuntime from 'VestRuntime';

export type IsolateKey = Nullable<string>;

export type TIsolate<P extends IsolatePayload = IsolatePayload> = {
  [IsolateKeys.AllowReorder]?: boolean;
  [IsolateKeys.Parent]: Nullable<TIsolate>;
  [IsolateKeys.Type]: string;
  [IsolateKeys.Keys]: Nullable<Record<string, TIsolate>>;
  [IsolateKeys.Data]: DataOnly<P>;
  children: Nullable<TIsolate[]>;
  key: IsolateKey;
  output: any;
} & UsedFeaturesOnly<P>;

type DataOnly<P extends IsolatePayload> = Omit<P, keyof IsolateFeatures>;
type UsedFeaturesOnly<P extends IsolatePayload> = Pick<
  P,
  keyof IsolateFeatures
>;

export class Isolate {
  static create<Payload extends IsolatePayload>(
    type: string,
    callback: CB,
    payload: Payload,
    key?: IsolateKey
  ): TIsolate<Payload> {
    const parent = VestRuntime.useIsolate();

    const newCreatedNode = IsolateMutator.setParent(
      baseIsolate(type, payload, key),
      parent
    );

    const nextIsolateChild = Reconciler.reconcile(newCreatedNode);

    const localHistoryNode = VestRuntime.useHistoryIsolateAtCurrentPosition();

    const output = Object.is(nextIsolateChild, newCreatedNode)
      ? useRunAsNew(localHistoryNode, newCreatedNode, callback)
      : nextIsolateChild.output;

    IsolateMutator.setParent(nextIsolateChild, parent);
    IsolateMutator.saveOutput(nextIsolateChild, output);
    VestRuntime.addNodeToHistory(nextIsolateChild);

    return nextIsolateChild as TIsolate<Payload>;
  }
}

/**
 * Creates a new child isolate context where the local history node is the current history node, thus advancing the history cursor.
 * Runs the callback function and returns its output.
 * @param localHistoryNode The local history node.
 * @param current The current isolate.
 * @param callback The callback function to execute.
 * @returns The output of the callback function.
 */
function useRunAsNew<Callback extends CB = CB>(
  localHistoryNode: Nullable<TIsolate>,
  current: TIsolate,
  callback: CB
): ReturnType<Callback> {
  const runtimeRoot = VestRuntime.useRuntimeRoot();

  // We're creating a new child isolate context where the local history node
  // is the current history node, thus advancing the history cursor.
  const output = VestRuntime.Run(
    {
      historyNode: localHistoryNode,
      runtimeNode: current,
      ...(!runtimeRoot && { runtimeRoot: current }),
    },
    () => callback(current)
  );

  current.output = output;
  return output;
}

function baseIsolate(
  type: string,
  payload: Record<string, any>,
  key: IsolateKey = null
): TIsolate {
  const { allowReorder, ...data } = payload;
  return {
    [IsolateKeys.AllowReorder]: allowReorder,
    [IsolateKeys.Keys]: null,
    [IsolateKeys.Parent]: null,
    [IsolateKeys.Type]: type,
    [IsolateKeys.Data]: data as IsolateData,
    children: null,
    key,
    output: null,
  };
}

type IsolateData = Record<string, any>;
type IsolatePayload = IsolateData & IsolateFeatures;
type IsolateFeatures = {
  [IsolateKeys.AllowReorder]?: boolean;
};
