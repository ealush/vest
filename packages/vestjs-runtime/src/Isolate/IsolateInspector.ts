import { Nullable, isNotNullish, isNullish } from 'vest-utils';

import { TIsolate } from 'Isolate';
import { IsolateKeys } from 'IsolateKeys';

export class IsolateInspector {
  static at(isolate: Nullable<TIsolate>, at: number): Nullable<TIsolate> {
    if (isNullish(isolate)) {
      return null;
    }
    return isolate.children?.[at] ?? null;
  }

  static cursor(isolate: Nullable<TIsolate>): number {
    if (isNullish(isolate)) {
      return 0;
    }
    return isolate.children?.length ?? 0;
  }

  static canReorder<I extends TIsolate>(isolate: Nullable<I>): boolean {
    if (isNullish(isolate)) {
      return false;
    }

    return IsolateInspector.allowsReorder(isolate.parent);
  }

  static allowsReorder<I extends Record<any, any>>(
    isolate: Nullable<I>
  ): boolean {
    return isolate?.allowReorder === true;
  }

  static usesKey(isolate: Nullable<TIsolate>): boolean {
    if (isNullish(isolate)) {
      return false;
    }
    return isNotNullish(isolate.key);
  }

  static getChildByKey(
    isolate: Nullable<TIsolate>,
    key: string
  ): Nullable<TIsolate> {
    if (isNullish(isolate)) {
      return null;
    }
    return isolate.keys?.[key] ?? null;
  }
}
