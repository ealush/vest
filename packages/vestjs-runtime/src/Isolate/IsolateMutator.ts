import { Nullable, invariant, isNullish } from 'vest-utils';

import { TIsolate } from 'Isolate';

export class IsolateMutator {
  static setParent(isolate: TIsolate, parent: Nullable<TIsolate>): TIsolate {
    isolate.parent = parent;
    return isolate;
  }

  static saveOutput(isolate: TIsolate, output: any): TIsolate {
    isolate.output = output;
    return isolate;
  }

  static setKey(isolate: TIsolate, key: Nullable<string>): TIsolate {
    isolate.key = key;
    return isolate;
  }

  static addChild(isolate: TIsolate, child: TIsolate): void {
    invariant(isolate);

    isolate.children = isolate.children ?? [];

    isolate.children.push(child);
    IsolateMutator.setParent(child, isolate);
  }

  static removeChild(isolate: TIsolate, node: TIsolate): void {
    isolate.children =
      isolate.children?.filter(child => child !== node) ?? null;
  }

  static addChildKey(isolate: TIsolate, key: string, node: TIsolate): void {
    invariant(isolate);

    isolate.keys = isolate.keys ?? {};

    isolate.keys[key] = node;
  }

  static slice(isolate: TIsolate, at: number): void {
    if (isNullish(isolate.children)) {
      return;
    }
    isolate.children.length = at;
  }

  static setData(isolate: TIsolate, data: any): void {
    isolate.data = data;
  }
}
