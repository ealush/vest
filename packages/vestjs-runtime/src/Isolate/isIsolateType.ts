import { Isolate } from 'Isolate';

export function isIsolateType(node: Isolate, type: string): node is Isolate {
  return node.type === type;
}
