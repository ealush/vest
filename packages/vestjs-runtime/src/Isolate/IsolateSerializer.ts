import { ErrorStrings } from 'ErrorStrings';
import {
  Nullable,
  hasOwnProperty,
  invariant,
  isNullish,
  isStringValue,
  text,
} from 'vest-utils';

import { TIsolate } from 'Isolate';
import { IsolateKeys, KeyToMinified, MinifiedToKey } from 'IsolateKeys';
import { IsolateMutator } from 'IsolateMutator';

export class IsolateSerializer {
  // eslint-disable-next-line max-statements, complexity
  static deserialize(node: Record<string, any> | TIsolate): TIsolate {
    // the  assumption is that the tree is built correctly,
    // but the children are missing the parent property to
    // avoid circular references during serialization.
    // in the same way, the parents are missing the `keys` property
    // to avoid circular references during serialization.
    // we need to rebuild the tree and add back the parent property to the children
    // and the keys property to the parents.

    const root = isStringValue(node)
      ? JSON.parse(node)
      : ({ ...node } as TIsolate);

    IsolateSerializer.validateIsolate(root);

    const queue = [root];

    while (queue.length) {
      const current = { ...queue.shift() } as TIsolate;

      const children = IsolateSerializer.getChildren(current);

      for (const key in MinifiedToKey) {
        if (hasOwnProperty(current, key)) {
          current[MinifiedToKey[key]] = current[key];
          delete current[key];
        }
      }

      if (!children) {
        continue;
      }

      for (const child of children) {
        IsolateMutator.setParent(child, current);
        queue.push(child);

        const key = child.key;

        if (key) {
          current.keys = current.keys ?? {};
          current.keys[key] = child;
        }
      }
    }

    return root;
  }

  static serialize(isolate: Nullable<TIsolate>): string {
    if (isNullish(isolate)) {
      return '';
    }

    return JSON.stringify(transformIsolate(isolate));
  }

  static getChildren(node: TIsolate): Nullable<TIsolate[]> {
    return node.children ? [...node.children] : null;
  }

  static validateIsolate(node: Record<string, any> | TIsolate): void {
    invariant(
      hasOwnProperty(node, IsolateKeys.Type),
      text(ErrorStrings.IVALID_ISOLATE_CANNOT_PARSE)
    );
  }
}

// eslint-disable-next-line complexity, max-statements
function transformIsolate(isolate: TIsolate): Record<string, any> {
  const next: Record<string, any> = {};

  if (isolate.children) {
    next.children = isolate.children.map(transformIsolate);
  }

  for (const key in isolate) {
    if (key === 'children') {
      continue;
    }

    if (isKeyExcluededFromDump(key)) {
      continue;
    }

    if (isNullish(isolate[key as keyof TIsolate])) {
      continue;
    }

    if (hasOwnProperty(KeyToMinified, key)) {
      next[KeyToMinified[key]] = isolate[key];
    }
  }

  return next;
}

function isKeyExcluededFromDump(key: string): boolean {
  return [IsolateKeys.Parent, IsolateKeys.Keys].includes(key as IsolateKeys);
}
