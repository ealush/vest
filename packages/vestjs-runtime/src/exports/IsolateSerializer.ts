import { ErrorStrings } from 'ErrorStrings';
import {
  Nullable,
  hasOwnProperty,
  invariant,
  isNotNullish,
  isNullish,
  isStringValue,
  text,
} from 'vest-utils';

import { TIsolate } from 'Isolate';
import { IsolateKeys, KeyToMinified, MinifiedToKey } from 'IsolateKeys';
import { IsolateMutator } from 'IsolateMutator';

export class IsolateSerializer {
  // eslint-disable-next-line max-statements, complexity
  static deserialize(node: Record<string, any> | TIsolate | string): TIsolate {
    // the  assumption is that the tree is built correctly,
    // but the children are missing the parent property to
    // avoid circular references during serialization.
    // in the same way, the parents are missing the `keys` property
    // to avoid circular references during serialization.
    // we need to rebuild the tree and add back the parent property to the children
    // and the keys property to the parents.

    const root = isStringValue(node)
      ? JSON.parse(node)
      : ({ ...node } as TIsolate);

    IsolateSerializer.validateIsolate(root);

    const queue = [root];

    while (queue.length) {
      const current = queue.shift();

      const children = IsolateSerializer.getChildren(current);

      for (const key in MinifiedToKey) {
        const value = current[key];
        if (isNotNullish(value)) {
          current[MinifiedToKey[key]] = value;
          delete current[key];
        }
      }

      if (!children) {
        continue;
      }

      current.children = children.map(child => {
        const nextChild = { ...child };

        IsolateMutator.setParent(nextChild, current);
        queue.push(nextChild);

        const key = nextChild.key;

        if (key) {
          current.keys = current.keys ?? {};
          current.keys[key] = nextChild;
        }

        return nextChild;
      });
    }

    return root as TIsolate;
  }

  static serialize(isolate: Nullable<TIsolate>): string {
    if (isNullish(isolate)) {
      return '';
    }

    return JSON.stringify(transformIsolate(isolate));
  }

  static getChildren(node: TIsolate): Nullable<TIsolate[]> {
    return node.children ? [...node.children] : null;
  }

  static validateIsolate(node: Record<string, any> | TIsolate): void {
    invariant(
      hasOwnProperty(node, IsolateKeys.Type) ||
        hasOwnProperty(node, KeyToMinified[IsolateKeys.Type]),
      text(ErrorStrings.IVALID_ISOLATE_CANNOT_PARSE)
    );
  }
}

// eslint-disable-next-line max-statements, complexity
function transformIsolate(isolate: TIsolate): Record<string, any> {
  const next: Record<string, any> = {};

  if (isolate.children) {
    next.children = isolate.children.map(transformIsolate);
  }

  for (const key in isolate) {
    if (key === 'children') {
      continue;
    }

    if (isKeyExcluededFromDump(key)) {
      continue;
    }
    const value = isolate[key as keyof TIsolate];

    if (isNullish(value)) {
      continue;
    }

    if (hasOwnProperty(KeyToMinified, key)) {
      next[KeyToMinified[key]] = value;
    } else {
      next[key] = value;
    }
  }

  return next;
}

function isKeyExcluededFromDump(key: string): boolean {
  return [IsolateKeys.Parent, IsolateKeys.Keys].includes(key as IsolateKeys);
}
