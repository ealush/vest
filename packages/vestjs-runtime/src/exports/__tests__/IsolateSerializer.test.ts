import { ser } from 'vest/src/testUtils/suiteDummy';

import { Isolate, TIsolate } from 'Isolate';
import { IsolateSerializer } from 'IsolateSerializer';
import { VestRuntime } from 'vestjs-runtime';

describe('IsolateSerializer', () => {
  describe('serialize', () => {
    it('Should produce serialized dump', () => {
      const { serialized } = createSerialized();

      expect(serialized).toMatchSnapshot();
    });
  });

  describe('deserialize', () => {
    it('Should fully inflate the tree', () => {
      const { root, serialized } = createSerialized();

      const inflated = IsolateSerializer.deserialize(serialized);

      expect(inflated).toMatchInlineSnapshot(`
        {
          "$type": "URoot",
          "children": [
            {
              "$type": "UChild_1",
              "data": {
                "some_data": true,
              },
              "parent": [Circular],
            },
            {
              "$type": "UChild_2",
              "data": {},
              "parent": [Circular],
            },
            {
              "$type": "UChild_3",
              "data": {},
              "parent": [Circular],
            },
          ],
          "data": {
            "some_data": true,
          },
        }
      `);
    });
  });
});

function withRunTime<T>(fn: CB<T>) {
  return VestRuntime.Run(VestRuntime.createRef({}), () => {
    return fn();
  });
}

function createSerialized() {
  let serialized: string, root: TIsolate;

  withRunTime(() => {
    root = Isolate.create(
      'URoot',
      () => {
        Isolate.create('UChild_1', () => {}, { some_data: true });
        Isolate.create('UChild_2', () => {});
        Isolate.create('UChild_3', () => {});
      },
      {
        some_data: true,
      }
    );

    serialized = IsolateSerializer.serialize(root);
  });

  // @ts-ignore
  return { root, serialized };
}
