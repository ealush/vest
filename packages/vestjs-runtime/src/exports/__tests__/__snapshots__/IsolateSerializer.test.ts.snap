// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IsolateSerializer serialize Should produce serialized dump 1`] = `"{"children":[{"$":"UChild_1","D":{"some_data":true},"S":"DONE"},{"$":"UChild_2","D":{},"S":"DONE"},{"$":"UChild_3","D":{},"S":"DONE"}],"$":"URoot","D":{"some_data":true},"S":"DONE"}"`;
